xblockutils/studio_editable.py

```
@@ -0,0 +1,295 @@
+                break
+        if "type" not in info:
+            raise NotImplementedError("StudioEditableXBlockMixin currently only supports fields derived from JSONField")
+        if info["type"] in ("list", "set"):
+            info["value"] = [json.dumps(val) for val in info["value"]]
+            if info["default"] is None:
+                info["default"] = []
+            info["default"] = json.dumps(info["default"])
+        if info["type"] == "generic":
+            # Convert value to JSON string if we're treating this field generically:
+            info["value"] = json.dumps(info["value"])
+            info["default"] = json.dumps(info["default"])
+        if 'values_provider' in field.runtime_options:
+            values = field.runtime_options["values_provider"](self)
+        else:
+            values = field.values
+        if values and not isinstance(field, Boolean):
+            info['has_values'] = True
+            # This field has only a limited number of pre-defined options.
+            # Protip: when defining the field, values= can be a callable.
+            if isinstance(field.values, dict) and isinstance(field, (Float, Integer)):
+                # e.g. {"min": 0 , "max": 10, "step": .1}
+                info["min"] = field.values["min"]
+                info["max"] = field.values["max"]
+                info["step"] = field.values["step"]
+            else:
+                # e.g. [1, 2, 3] or [ {"display_name": "Always", "value": "always"}, {...}, ... ]
+                if not isinstance(values[0], dict) or "display_name" not in values[0]:
+                    values = [{"display_name": val, "value": val} for val in values]
+                info['values'] = values
+        if info["type"] in ("list", "set") and field.runtime_options.get('list_values_provider'):
+            list_values = field.runtime_options['list_values_provider'](self)
+            # list_values must be a list of values or {"display_name": x, "value": y} objects
```

I'm wary of statements like this - if incoming data must have a particular form, I'd rather see explicit lines that asserts the form of the data and fails hard if the data's form is incorrect. That makes understanding errors _much_ easier to catch.